File: burger.cpp
#include <iostream>
#include <cmath>
#include <fstream>
#include <omp.h>
#include "burger.hpp"

#define EXPLICIT 1
#define IMPLICIT 2
#define NGHOST 2

int main(int argc, char *argv[]) {

    // struct to hold all PDE integration parameters
    struct pdeParams params;

    int scheme;

    if(argc==5) {
        params.nx = atoi(argv[1]);
        params.nwg = params.nx + NGHOST;
        params.nsteps = atoi(argv[2]);
        params.v = atof(argv[3]);
        scheme = atoi(argv[4]);
    } else {
        std::cout << "Please supply 4 arguments specifying nx, nsteps, v,";
        std::cout << "and which scheme: 1 for explicit, 2 for implicit \n";
        exit(1);
    }

    // specify length of domain and the final time
    double L = 2*M_PI;
    double tf = 5;

    params.dx = L/params.nx;
    params.dt = tf/params.nsteps;


    Vec u(params.nwg);

    // apply initial conditions
    init(u.data(),params);

    if(scheme == EXPLICIT) {
        // explicit, upwind scheme
        expl(u.data(),params);
    } else if(scheme == IMPLICIT) {
        implicit(u.data(),params);
    }

    // write to file
    std::ofstream arr_file("u.dat");
    if(arr_file.is_open()) {
        for(int i = 0; i < params.nx; ++i) {
            arr_file << params.dx*i << "\t" << u[i] << "\n";
        }
    } else {
        std::cout << "Can't write to file u.dat \n";
        exit(1);
    }

    return 0;
}

int init(double* u, struct pdeParams params) {
    // Writes initial condition to state vector u
    int nwg = params.nwg;
    double dx = params.dx;

    for(int i = 0; i < nwg; ++i) {
        double x = dx*i;
        u[i] = sin(x)*exp(-(pow(x-M_PI,2)));
    }

    return 0;
}

int implicit(double* u, struct pdeParams params) {

    // unpack struct for convenient naming
    int nwg = params.nwg;
    int nx = params.nx;
    int nsteps = params.nsteps;
    double v = params.v;
    double dx = params.dx;
    double dt = params.dt;
    
    Vec jac(nwg*nwg);
    Vec u0(nwg);
    Vec a(nwg);

    // rhs value placeholder
    double f;

    // number of newton iterations per step
    int iter = 100;

    // gameplan: do an fe step as a first guess, then do newton iterations to find next timestep
    for(int i = 0; i < nsteps; ++i) {
        // Periodic Boundary Conditions
        u[0] = u[nwg-4];
        u[1] = u[nwg-3];
        u[nwg-2] = u[2];
        u[nwg-1] = u[3];

        // FE Step
        for(int j = 1; j < nwg-1; ++j) {
            f = v/(dx*dx)*(u[j+1]-2*u[j]+u[j-1])-u[j]/(2*dx)*(u[j+1]-u[j-1]);
            u0[j] = u[j] + dt*f;
        }

        // calculate jacobian based on guess u0
        jacobian(jac.data(),u0.data(),params);
        
        // Newton iteration
        for(int k = 1; k < iter; ++k) {
            // gameplan: evaluate a
            // invert jac
            //
            for(int j = 1; j < nwg-1; ++j) {
            }
        }

    }
    

    return 0;
}

int jacobian(double* jac, const double* u, struct pdeParams params) {
    // given state vector u and PDE parameters params, determine
    // jacobian of nonlinear system


    int nx = params.nx;
    int nwg = params.nwg;
    double dx = params.dx;
    double dt = params.dt;
    double v = params.v;

    // periodicity automatically enforced with ghost points
    for(int i = 1; i < nwg-1; ++i) {
        jac[nwg*i+i] = 1 + dt*(1/(2*dx)*(u[i+1]-u[i-1])+2*v/(dx*dx));   // main diagonal
        jac[nwg*i+i-1] = dt*(1/(2*dx)*(-u[i])-v/(dx*dx));              // sub diagonal
        jac[nwg*i+i+1] = dt*(1/(2*dx)*(u[i])-v/(dx*dx));              // super diagonal
    }

    return 0;
}

int expl(double* u, struct pdeParams params) {
    // implementation of explicit scheme using 
    // 2-point upwinded differencing and forward euler

    double v = params.v;
    double dx = params.dx;
    double dt = params.dt;

    int i;
    int j;

    double f;
    int coeff;

    // main loop
    // could parallelize this easily but it's not worth it
    for(i = 0; i < params.nsteps; ++i){
        // perodic boundary conditions with ghost points
        u[0] = u[params.nwg-4];
        u[1] = u[params.nwg-3];
        u[params.nwg-2] = u[2];
        u[params.nwg-1] = u[3];
        for (j = 1; j < params.nwg-2; ++j) {
            coeff = (u[j]>0); // if u[j] > 0, then use backward difference else forward difference
            f = v/(dx*dx)*(u[j+1]-2*u[j]+u[j-1])-u[j]/dx*(u[j-coeff+1]-u[j-coeff]);
            u[j] += dt*f;
        }
    }
    return 0;
}

File: src/burger.jl
using LinearAlgebra

function main()

    # Set PDE boundaries
    L::Float64 = 2π
    Tf::Float64 = 5

    # PDE parameters
    nx::Int64 = 5
    nt::Int64 = 5000
    Δx::Float64 = L/nx
    Δt::Float64 = Tf/nt
    v::Float64 = 0.01

    x = [Δx*i for i = 1:nx]

    u = [sin(x[i])*exp(-(x[i]-π)^2) for i = 1:nx]

    T = upwind(u,nx,Δx,Δt,v)

    # explicit(u,nx,nt,Δx,Δt,v)

    return x,u,T

end

function explicit(u::Array{Float64}, nx::Int64, nt::Int64, Δx::Float64,Δt::Float64, v::Float64)

    # Main loop
    for i=1:nt
        T = upwind(u,nx,Δx,Δt,v)
        u += Δt.*(T*u)
    end

    return 0

end

function upwind(u::Array{Float64}, nx::Int64, Δx::Float64, Δt::Float64, v::Float64)

    # 1st Order Finite Difference
    # Main diagonal of the operator matrix
    # d1 = [if u[i] > 0 1 else (-1) end for i = 1:nx]
    # l1 = [if u[i] > 0 (-1) else 0 end for i = 2:nx]
    # u1 = [if u[i] > 0 0 else 1 end for i = 1:nx-1]


    # Pure Advection Diffusion
    d1 = [1. for i = 1:nx]
    l1 = [-1. for i = 1:nx-1]
    u1 = [0.0 for i = 1:nx-1]

    # Periodic Boundary Conditions
    T1 = [0 for i = 1:nx, j = 1:nx] + Tridiagonal(l1,d1,u1)
    T1[1,nx] = u[1] > 0 ? (-1) : 0
    T1[nx,1] = u[nx] > 0 ? 0 : 1

    # 2nd Order Finite Difference
    d2 = [-2 for i = 1:nx]
    ul2 = [1 for i = 1:nx-1]
    T2 = [0 for i = 1:nx, j = 1:nx] + Tridiagonal(ul2,d2,ul2)
    T2[1,nx] = 1
    T2[nx,1] = 1

    T = (T2.*(v/Δx^2)-T1./Δx)

    return T1

end

main()

